scripts/service.py

```diff
@@ -0,0 +1,286 @@
+#!/usr/bin/env python
+#
+# Copyright (C) Citrix Systems Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published
+# by the Free Software Foundation; version 2.1 only. #
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# This script can be used as both server as client. For server mode, it is expected to
+# be started as daemon, wait on some named pipe depending on input, and upon wake up
+# (client side writes the pipe), read out all content from the pipe, and execute a
+# pre-configed (through input parameter) scripts. This can avoid input storm so if multiple
+# client side invoke, server side just do one invoke of the executable (assume the 
+# script do not need to be executed multiple times in case of multiple trigger). For
+# client side, it is just simple logic to open the same service pipe as server, and write
+# something to wake up the server to do the work.
+#
+# There are multiple input parameters,
+#    -s/--service:  this is the service name for both server and client, supposing a short name
+#    -m/--mode: this is the mode specify 'client' or 'server'
+#    -e/--executable: this is the executable script to be invoked upon server wake up, 
+#                     only needed for server logic. Arguments have to be included along with
+#                     script name as one string.
+#
+# Example:
+#    server:
+#       service.py -s usbscan -m server -e /opt/xensource/libexec/usb_change
+#
+#    client:
+#        service.py -s usbscan -m client
+#
+
+import argparse
+import os
+import select
+import subprocess
+import logging
+import xcp.logger as logger
+import errno
+import shutil
+import stat
+import sys
+
+BUFFER_SIZE = 2048
+
+class ServiceSkeleton(object):
+    """ base class for both client and server, maily as utilities """
+
+    PIPE_BASE_PATH = '/var/lib/misc/'
```

Comment:
Put the pipes in /var/run/udevworker or something similar under /var/run, would be my advice. Since /var/run is a tmpfs, you are guaranteed not to have old rubbish lying around after a reboot, which makes everything a bit more predictable.

Reply:
OK, I originally considered, but did not do just to avoid complicated init file/folder sequence.